daemon/disk_usage.go

```diff
@@ -5,46 +5,60 @@ import (
 	"fmt"
 	"sync/atomic"
 
+	"github.com/docker/docker/api/server/router/system"
 	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/api/types/filters"
 )
 
 // SystemDiskUsage returns information about the daemon data disk usage
-func (daemon *Daemon) SystemDiskUsage(ctx context.Context) (*types.DiskUsage, error) {
+func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsageOptions) (*types.DiskUsage, error) {
```

Comment:
I looked into possibility of using functional options here, but the issue is that it would require us to define the exact same struct of `bool`s in `api/server/router/system` in order to make things like `WithContainers()` work, since internally they would have to set a boolean field. Of course the internal struct could be abstracted, and getters could be exposed for the implementation to use, but overall, I do not see benefits of using this approach in this particular case, where a simple struct of bools suffices.

Reply:
Yeah, I guess if we want those, one option _could be_ to make the signature for the option `func(*types.DiskUsage) error`, and the functions that are passed would do the actual collecting, e.g. (quick blurb)

```go
func WithContainerUsage() UsageOpt
	return func(usage *types.DiskUsage) error {
		// Retrieve container list
		containers, err = daemon.Containers(&types.ContainerListOptions{
			Size: true,
			All:  true,
		})
		if err != nil {
			return nil, fmt.Errorf("failed to retrieve container list: %v", err)
		}
		usage.Containers = containers
		return nil
	}
}
```

Reply:
Yes, mostly "theoretical", I don't think there's much benefit (if any) over the current approach. If we _would_ have to define it, we can't do so in the daemon (as the API should not import the daemon), so then the `daemon.Containers` would have to be `<backend>.Containers`, and the backend passed in.